inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/hook/HookManager.java

```diff
@@ -0,0 +1,221 @@
+
+    @Autowired
+    private BatchJobExecutorService executor;
+
+    @Autowired
+    private InspectitEnvironment env;
+
+    @Autowired
+    private Instrumentation instrumentation;
+
+    @Autowired
+    private InstrumentationConfigurationResolver configResolver;
+
+    @Autowired
+    private SelfMonitoringService selfMonitoring;
+
+    @Autowired
+    private MethodHookGenerator hookGenerator;
+
+    private LoadingCache<Class<?>, ConcurrentHashMap<String, MethodHook>> hooks = CacheBuilder.newBuilder().weakKeys().build(
+            new CacheLoader<Class<?>, ConcurrentHashMap<String, MethodHook>>() {
+                @Override
+                public ConcurrentHashMap<String, MethodHook> load(Class<?> key) throws Exception {
+                    return new ConcurrentHashMap<>();
+                }
+            });
+
+
+    /**
+     * The set of classes which might need hook updates.
+     * This service works through this set in batches.
+     */
+    @VisibleForTesting
+    Cache<Class<?>, Boolean> pendingClasses =
+            CacheBuilder.newBuilder().weakKeys().build();
+
+
+    private BatchJobExecutorService.BatchJob<Integer> classInstrumentationJob;
+
+    @PostConstruct
+    void init() {
+        Instances.hookManager = this::getHook;
+
+        InternalSettings conf = env.getCurrentConfig().getInstrumentation().getInternal();
+        Duration delay = conf.getInterBatchDelay();
+
+        classInstrumentationJob = executor.startJob(this::checkClassesForConfigurationUpdates, conf.getClassConfigurationCheckBatchSize(), delay, delay);
+    }
+
+    @PreDestroy
+    void destroy() {
+        Instances.hookManager = NoopHookManager.INSTANCE;
+        classInstrumentationJob.cancel();
+    }
+
+
+    private IMethodHook getHook(Class<?> clazz, String methodSignature) {
+        try {
+            val hook = hooks.get(clazz).get(methodSignature);
+            return hook == null ? NoopMethodHook.INSTANCE : hook;
+        } catch (ExecutionException e) {
+            //Should in theory never occur, as we only call new ConcurrentHashMap() in the cache loader
+            log.error("Error in cache", e);
+            return NoopMethodHook.INSTANCE;
+        }
+    }
+
+
+    @Override
+    public void onNewClassesDiscovered(Set<Class<?>> newClasses) {
+        for (Class<?> clazz : newClasses) {
+            pendingClasses.put(clazz, Boolean.TRUE);
+        }
+        selfMonitorQueueSize();
+    }
+
+    @EventListener
+    private void configEventListener(InspectitConfigChangedEvent ev) {
+        InternalSettings newInternal = ev.getNewConfig().getInstrumentation().getInternal();
+        classInstrumentationJob.setBatchSizes(newInternal.getClassConfigurationCheckBatchSize());
+        classInstrumentationJob.setInterBatchDelay(newInternal.getInterBatchDelay());
+    }
+
+    @EventListener
+    private void instrumentationConfigEventListener(InstrumentationConfigurationChangedEvent ev) {
```

Comment:
simply call onNewClassesDiscovered(`instrumentation.getAllLoadedClasses())`? Change `Set` to be a `Collection` in the signature, or?

Reply:
I will add a comment to both methods refererring to each other